File: 2023/08.07/송기훈/구슬탈출3.java
package baekjoon.gold.one;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Objects;
import java.util.Queue;
import java.util.StringTokenizer;

public class 구슬탈출3 {

    static class Point {
        int rY, rX;
        int bY, bX;
        int depth;
        String dir;

        public Point(int rY, int rX, int bY, int bX, int depth, String dir) {
            this.rY = rY;
            this.rX = rX;
            this.bY = bY;
            this.bX = bX;
            this.depth = depth;
            this.dir = dir;
        }

    }

    static int N, M;
    static int rY, rX;
    static int bY, bX;
    static int hY, hX;
    static char[][] matrix;

    // 하 우 상 좌
    static int[] dY = {1, 0, -1, 0};
    static int[] dX = {0, 1, 0, -1};
    static String[] dir = {"D", "R", "U", "L"};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());

        matrix = new char[N][M];
        for (int i = 0; i < N; i++) {
            String inputString = br.readLine();
            for (int j = 0; j < M; j++) {
                char temp = inputString.charAt(j);

                if (temp == 'R') {
                    rY = i;
                    rX = j;
                } else if (temp == 'B') {
                    bY = i;
                    bX = j;
                } else if (temp == 'O') {
                    hY = i;
                    hX = j;
                }

                matrix[i][j] = temp;
            }
        }

        bfs(new Point(rY, rX, bY, bX, 0, ""));

    }

    static void bfs(Point point) {
        Queue<Point> queue = new LinkedList<>();
        queue.offer(point);
        boolean[][][][] visited = new boolean[N][M][N][M];
        visited[point.rY][point.rX][point.bY][point.bX] = true;

        while (!queue.isEmpty()) {
            Point p = queue.poll();

            // 10 넘어가면 -1 출력
            if (p.depth > 10) {
                System.out.println(-1);
                return;
            }

            // 파란 구슬이 빠지면 안됨
            if (matrix[p.bY][p.bX] == 'O') {
                continue;
            }

            // 빨간 구슬이 O, 파란 구슬이 O가 아니면 성공
            if (matrix[p.rY][p.rX] == 'O' && matrix[p.bY][p.bX] != 'O') {
                System.out.println(p.depth);
                System.out.println(p.dir);
                return;
            }

            for (int i = 0; i < 4; i++) {
                int rY = p.rY;
                int rX = p.rX;

                // 빨간 구슬을 한 쪽 방향으로 이동
                while (true) {
                    if (rY + dY[i] >= 0 && rX + dX[i] >= 0) {
                        rY += dY[i];
                        rX += dX[i];

                        // 벽에 닿았을 경우 한 칸 빼주자
                        if (matrix[rY][rX] == '#') {
                            rY -= dY[i];
                            rX -= dX[i];
                            break;
                        }

                        // O에 도착
                        if (matrix[rY][rX] == 'O') {
                            break;
                        }
                    }
                }

                int bY = p.bY;
                int bX = p.bX;

                // 파란 구슬을 한 쪽 방향으로 이동
                while (true) {
                    if (bY + dY[i] >= 0 && bX + dX[i] >= 0) {
                        bY += dY[i];
                        bX += dX[i];

                        // 벽에 닿았을 경우 한 칸 빼주자
                        if (matrix[bY][bX] == '#') {
                            bY -= dY[i];
                            bX -= dX[i];
                            break;
                        }

                        // O에 도착
                        if (matrix[bY][bX] == 'O') {
                            break;
                        }
                    }
                }

                // 빨간 구슬, 파란 구슬이 같은 곳이면?
                // #...RB# --L--> #RB...#
                if (rY == bY && rX == bX && matrix[rY][rX] != 'O') {
                    int rDiff = Math.abs(rY - p.rY) + Math.abs(rX - p.rX);
                    int bDiff = Math.abs(bY - p.bY) + Math.abs(bX - p.bX);

                    // 벽으로부터 더 먼 것을 한 칸 빼준다
                    if (rDiff > bDiff) {
                        rY -= dY[i];
                        rX -= dX[i];
                    } else {
                        bY -= dY[i];
                        bX -= dX[i];
                    }
                }

                if (!visited[rY][rX][bY][bX]) {
                    visited[rY][rX][bY][bX] = true;
                    queue.offer(new Point(rY, rX, bY, bX, p.depth + 1, p.dir + dir[i]));
                }
            }
        }
        System.out.println(-1);
    }

}

File: 2023/08.07/조수연/구슬_탈출_3.kt
import java.awt.Point
import java.io.*
import java.util.*
import kotlin.math.abs

lateinit var board: Array<CharArray>
lateinit var visited: Array<Array<Array<BooleanArray>>>
var sequence = ""
val dir = arrayOf(intArrayOf(-1, 0), intArrayOf(1, 0), intArrayOf(0, -1), intArrayOf(0, 1))

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    val (n, m) = readLine().trim().split(" ").map { it.toInt() }
    board = Array(n) { CharArray(m) }
    visited = Array(n) { Array(m) { Array(n) { BooleanArray(m) } } }
    val red = Point()
    val blue = Point()

    repeat(n) { i ->
        val info = readLine()
        repeat(m) { j ->
            board[i][j] = info[j]
            if (board[i][j] == 'R') {
                red.x = j
                red.y = i
            } else if (board[i][j] == 'B') {
                blue.x = j
                blue.y = i
            }
        }
    }

    val cnt = bfs(red, blue)
    bw.write("$cnt\n$sequence")

    bw.flush()
    bw.close()
}

fun bfs(red: Point, blue: Point): Int {
    val q: Queue<Triple<Point, Point, Pair<Int, String>>> = LinkedList()
    val direction = charArrayOf('U','D','L','R')
    
    visited[red.y][red.x][blue.y][blue.x] = true
    q.add(Triple(red, blue, Pair(0, "")))

    while (q.isNotEmpty()) {

        val info = q.poll()
        val r = info.first
        val b = info.second
        val s = info.third

        if (s.first > 10) return -1

        if (board[b.y][b.x] == 'O') continue // 파란 공이 구멍에 빠지면 안된다.

        if (board[r.y][r.x] == 'O' && board[b.y][b.x] != 'O') { // 빨간 공만 구멍에 빠졌을 경우 종료
            sequence = s.second
            return s.first
        }

        for (i in 0 until 4) {
            var nrx = r.x
            var nry = r.y

            while (true) {
                if (nrx + dir[i][1] >= 0 && nry + dir[i][0] >= 0) {
                    nrx += dir[i][1]
                    nry += dir[i][0]

                    if (board[nry][nrx] == '#' || board[nry][nrx] == 'O') break
                }
            }

            if (board[nry][nrx] == '#') {
                nrx -= dir[i][1]
                nry -= dir[i][0]
            }

            var nbx = b.x
            var nby = b.y

            while (true) {
                if (nbx + dir[i][1] >= 0 && nby + dir[i][0] >= 0) {
                    nbx += dir[i][1]
                    nby += dir[i][0]

                    if (board[nby][nbx] == '#' || board[nby][nbx] == 'O') break
                }
            }

            if (board[nby][nbx] == '#') {
                nbx -= dir[i][1]
                nby -= dir[i][0]
            }

            if (nry == nby && nrx == nbx && board[nry][nrx] != 'O') { // 같은 위치인 경우
                val rDis = abs(nry - r.y) + abs(nrx - r.x)
                val bDis = abs(nby - b.y) + abs(nbx - b.x)

                if (rDis > bDis) { // 파란 공은 못 움직이는 상황에서 빨간 공이랑 만난 상황
                    nrx -= dir[i][1]
                    nry -= dir[i][0]
                } else {
                    nbx -= dir[i][1]
                    nby -= dir[i][0]
                }
            }

            // 두 공이 이전의 위치와 같은지 체크
            if (!visited[nry][nrx][nby][nbx]) {
                visited[nry][nrx][nby][nbx] = true
                q.add(Triple(Point(nrx, nry), Point(nbx, nby), Pair(s.first + 1, s.second + direction[i])))
            }
        }
    }
    return -1
}